src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1819,14 +1820,61 @@ internal void LogExecutingCompleted(NpgsqlConnector connector, bool executing)
             }
         }
 
-        object[] ParametersDbNullAsString(NpgsqlBatchCommand c)
+        static object[] GetParametersForLogging(NpgsqlBatchCommand c)
         {
             var positionalParameters = c.CurrentParametersReadOnly;
             var parameters = new object[positionalParameters.Count];
             for (var i = 0; i < positionalParameters.Count; i++)
-                parameters[i] = positionalParameters[i].Value == DBNull.Value ? "NULL" : positionalParameters[i].Value!;
+            {
+                var value = positionalParameters[i].Value;
+                object? stringValue = null;
+                parameters[i] = TryFormatArgumentIfNullOrEnumerable(value, ref stringValue)
+                    ? stringValue
+                    : value!;
+            }
+
             return parameters;
         }
+
+        static bool TryFormatArgumentIfNullOrEnumerable([NotNullWhen(true)] object? value, [NotNullWhen(true)] ref object? stringValue)
+        {
+            const string DbNullValue = "NULL";
+            const string ClrNullValue = "(null)";
+
+            if (value is null)
+            {
+                stringValue = ClrNullValue;
+                return true;
+            }
+
+            if (value == DBNull.Value)
+            {
+                stringValue = DbNullValue;
+                return true;
+            }
+
+            if (value is not string && value is IEnumerable enumerable)
+            {
+                var vsb = new StringBuilder(256);
```

Comment:
One problem with writing arrays is that they can be quite long (with byte arrays it quickly gets out of hand), which is a problem with both string allocation and trying to read all that is logged. I propose that we only write values with some limit (let's say the first 10, afterwards it's just `...` to indicate that there were more). Usually stuff like this handled at the logger itself, since users can change the upper limit + it doesn't have to allocate an additional string and instead just directly write to stream.